File: MediTimer/Domain/Services/AudioSessionCoordinatorProtocol.swift
//
//  AudioSessionCoordinatorProtocol.swift
//  MediTimer
//
//  Domain Service Protocol - Audio Session Coordinator
//

import Combine
import Foundation

/// Audio source identifier for coordination
enum AudioSource: String, Equatable {
    case timer
    case guidedMeditation
}

/// Centralized coordinator for audio session management
///
/// This coordinator ensures that only one audio source is active at a time,
/// preventing conflicts between timer background audio and guided meditation playback.
///
/// Usage:
/// ```swift
/// // Register callback for when another source takes over
/// coordinator.registerConflictHandler(for: .timer) {
///     // Stop audio playback
/// }
///
/// // Request audio session before playing
/// try coordinator.requestAudioSession(for: .timer)
///
/// // Release when done
/// coordinator.releaseAudioSession(for: .timer)
/// ```
protocol AudioSessionCoordinatorProtocol: AnyObject {
    /// Currently active audio source (nil if none)
    var activeSource: CurrentValueSubject<AudioSource?, Never> { get }

    /// Registers a conflict handler for the given audio source
    ///
    /// The handler will be called synchronously when another source requests the audio session.
    /// This allows the current source to cleanly stop its audio playback.
    ///
    /// - Parameters:
    ///   - source: The audio source registering the handler
    ///   - handler: Closure called when another source becomes active
    func registerConflictHandler(for source: AudioSource, handler: @escaping () -> Void)

    /// Requests exclusive use of the audio session
    ///
    /// If another source is currently active, its conflict handler will be called synchronously.
    ///
    /// - Parameter source: The audio source requesting the session
    /// - Returns: True if session was granted, false if request was denied
    /// - Throws: Audio session configuration errors
    func requestAudioSession(for source: AudioSource) throws -> Bool

    /// Releases the audio session for the given source
    ///
    /// If this source is currently active, the audio session will be deactivated
    /// to save energy.
    ///
    /// - Parameter source: The audio source releasing the session
    func releaseAudioSession(for source: AudioSource)

    /// Configures and activates the shared audio session
    ///
    /// - Throws: Audio session configuration errors
    func activateAudioSession() throws

    /// Deactivates the shared audio session
    ///
    /// Should only be called when no audio sources are active.
    func deactivateAudioSession()
}

File: android/app/src/main/kotlin/com/stillmoment/domain/services/AudioSessionCoordinatorProtocol.kt
package com.stillmoment.domain.services

import com.stillmoment.domain.models.AudioSource
import kotlinx.coroutines.flow.StateFlow

/**
 * Protocol for coordinating exclusive audio session access between features.
 *
 * Only one audio source (Timer or Guided Meditation) can be active at a time.
 * When a new source requests access, the current source is notified via its
 * registered conflict handler and must release the session.
 *
 * This mirrors the iOS AudioSessionCoordinatorProtocol for feature parity.
 */
interface AudioSessionCoordinatorProtocol {
    /**
     * The currently active audio source, or null if no source is active.
     */
    val activeSource: StateFlow<AudioSource?>

    /**
     * Register a conflict handler for a specific audio source.
     * The handler is invoked when another source requests the audio session.
     *
     * @param source The audio source registering the handler
     * @param handler The callback to invoke when a conflict occurs
     */
    fun registerConflictHandler(source: AudioSource, handler: () -> Unit)

    /**
     * Request exclusive audio session access for a source.
     *
     * If another source is currently active, its conflict handler is invoked
     * before granting access to the requesting source.
     *
     * @param source The audio source requesting access
     * @return true if access was granted
     */
    fun requestAudioSession(source: AudioSource): Boolean

    /**
     * Release the audio session for a source.
     *
     * Only releases if the specified source is currently active.
     *
     * @param source The audio source releasing the session
     */
    fun releaseAudioSession(source: AudioSource)
}
